### PR TITLE
Add bridge-messages pallet to pangolin parachain

### PR DESCRIPTION
**Pull request by furoxr**

### CODE REVIEW

**Review by hackfisher — COMMENTED** [review]
Config fee market in this pull request too.

**Comment by boundless-forest** [comment]
@furoxr Please update the  `pallet-bridge-messages` and use the laest fee-market pallet.  https://github.com/darwinia-network/darwinia-common/pull/1158

**Comment by acgxv** [comment]
@AsceticBear Please take a review.

**Review by hackfisher — APPROVED** [review]

File: runtime/crab-parachain/src/lib.rs [hunk 1]
```diff
@@ -349,6 +349,15 @@ sp_api::impl_runtime_apis! {
 		}
 	}
 
+	impl pallet_fee_market_rpc_runtime_api::FeeMarketApi<Block, Balance> for Runtime {
+		fn market_fee() -> Option<pallet_fee_market_rpc_runtime_api::Fee<Balance>> {
+			Default::default()
```
Comment (boundless-forest) [inline runtime/crab-parachain/src/lib.rs:354]: Curious why this API implementation is empty.

Comment (acgxv) [reply runtime/crab-parachain/src/lib.rs:354]: S2S bridge is not available for Crab*/Darwinia*.

File: runtime/pangolin-parachain/src/pallets/fee_market.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			))
+			.unique_saturated_into();
+
+		cmp::min(locked_collateral, slash_value)
+	}
+}
+
+frame_support::parameter_types! {
+	pub const FeeMarketPalletId: PalletId = PalletId(*b"da/feemk");
+	pub const TreasuryPalletId: PalletId = PalletId(*b"da/trsry");
```
Comment (acgxv) [inline runtime/pangolin-parachain/src/pallets/fee_market.rs:26]: IIUC, fee market doesn't need treasury pallet. Just an account here.

File: runtime/crab-parachain/src/lib.rs (hunk 1 above)
Comment (boundless-forest) [reply runtime/crab-parachain/src/lib.rs:354]: I am not aware of the reason~, can you explain more about this? The bridge-messages pallets added in this pr will be used in Crab<> Crab parachain. What's the relationship with Darwinia Chain here?

Comment (acgxv) [reply runtime/crab-parachain/src/lib.rs:354]: For Pangolin<>Pangolin Parachain. Leave empty for Crab*/Darwinia*.

File: runtime/pangolin-parachain/src/pallets/fee_market.rs (hunk 2 above)
Comment (boundless-forest) [reply runtime/pangolin-parachain/src/pallets/fee_market.rs:26]: fee market uses this id to generate the treasury account_id. Do you mean hardcode the treasury account_id in the fee market runtime config directly?

File: runtime/pangolin-parachain/src/bridge_messages/pangolin.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+use pallet_bridge_messages::EXPECTED_DEFAULT_MESSAGE_LENGTH;
+
+/// Message verifier for PangolinParachain -> Pangolin messages.
+pub type ToPangolinMessageVerifier = FromThisChainMessageVerifier<WithPangolinMessageBridge>;
+/// Message payload for PangolinParachain -> Pangolin messages.
+pub type ToPangolinMessagePayload = FromThisChainMessagePayload<WithPangolinMessageBridge>;
+
+/// Message payload for Pangolin -> PangolinParachain messages.
+pub type FromPangolinMessagePayload = FromBridgedChainMessagePayload<WithPangolinMessageBridge>;
+/// Call-dispatch based message dispatch for PangolinParachain -> Pangolin messages.
```
Comment (boundless-forest) [inline runtime/pangolin-parachain/src/bridge_messages/pangolin.rs]:
```suggestion
/// Call-dispatch based message dispatch for  Pangolin -> PangolinParachain messages.
```

**Comment by acgxv** [comment]
It seems I missed `FromThisChainMessageVerifier`

File: runtime/crab-parachain/src/lib.rs (hunk 1 above)
Comment (boundless-forest) [reply runtime/crab-parachain/src/lib.rs:354]: suggest adding some comments like `TOO: xxx`

File: runtime/pangolin-parachain/src/bridge_messages/pangolin.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+use sp_std::ops::RangeInclusive;
+// --- darwinia-network ---
+use crate::*;
+use bp_messages::{source_chain::TargetHeaderChain, target_chain::*, *};
+use bp_runtime::{Chain, ChainId, PANGOLIN_CHAIN_ID, PANGOLIN_PARACHAIN_CHAIN_ID};
+use bridge_runtime_common::messages::{self, source::*, target::*, *};
+use pallet_bridge_messages::EXPECTED_DEFAULT_MESSAGE_LENGTH;
+
+/// Message verifier for PangolinParachain -> Pangolin messages.
+pub type ToPangolinMessageVerifier = FromThisChainMessageVerifier<WithPangolinMessageBridge>;
```
Comment (boundless-forest) [inline runtime/pangolin-parachain/src/bridge_messages/pangolin.rs]: Compare with https://github.com/darwinia-network/darwinia-common/blob/main/node/runtime/pangoro/src/pangolin_messages.rs#L67. Loss generic param R, is that expected?

File: runtime/pangolin-parachain/src/pallets/fee_market.rs (hunk 2 above)
Comment (acgxv) [reply runtime/pangolin-parachain/src/pallets/fee_market.rs:26]: It's okay without treasury pallet.
Treasury account still will be created based on the id.

**Comment by acgxv** [comment]
@AsceticBear LGTM now.

**Review by boundless-forest — APPROVED** [review]
LGTM now.

**Pull request merged**